rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,143 @@
+# Integrate `cosign` with `kpack`
+
+## Problem
+
+The signing solution currently supported by `kpack` is
+[Notary v1](https://github.com/theupdateframework/notary). However, Notary
+presents itself with several different challenges, including:
+- It is unable to sign container image digests. The user must always specify a
+tag to be signed.
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures, using
+  either:
+  - `crane cp` to copy images and signatures separately.
+  - `cosign copy` to copy images along with their signatures from one registry
+    to another.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Implement a flow that generates the signature payload for the image, then
+  calculates its signature and pushes it either to the registry where the image
+  is located, using the same credentials that were used to push the image, or
+  [to the registry specified in the `COSIGN_REPOSITORY` environment variable](#key-generation-and-storage).
```

Comment:
How would the `COSIGN_REPOSITORY` environment variable be set?

Reply:
This would be an implementation detail. We could mount the secret into the build pod and consume it from there, or use other mechanisms to pass it in, depending on the preferred approach. `cosign` currently consumes it from the environment variable. However, I assume that there might be a flow where we don't necessarily have it set in the environment.

Reply:
Given the current code in cosign, which only supports this environment variable, kpack's code would need to set it before invoking the code that uploads signatures.

Reply:
The user will need to be able to specify the value for this. I think we previously discussed putting it in the Image's `cosign` section as a field?

Reply:
> The user will need to be able to specify the value for this. I think we previously discussed putting it in the Image's `cosign` section as a field?

We indeed talked about it, but given that there might be a situation where a user could want to push different signatures to different repositories, this field was added to the secret instead. I linked the section where we are talking about it. Happy to reorder sections if it helps reading! 🙂